x/incentives/keeper/gauge_asset.go

```diff
@@ -140,34 +142,92 @@ func (k Keeper) distributeToAssetGauge(ctx sdk.Context, gauge types.Gauge, currR
 	}
 
 	totalDistrCoins := sdk.NewCoins()
+	lockSumTimesRemainingEpochs := lockSum.MulRaw(int64(remainEpochs))
+
 	for _, lock := range locks {
 		distrCoins := sdk.Coins{}
 		for _, coin := range remainCoins {
 			// distribution amount = gauge_size * denom_lock_amount / (total_denom_lock_amount * remain_epochs)
-			denomLockAmt := lock.Coins.AmountOfNoDenomValidation(denom)
-			amt := coin.Amount.Mul(denomLockAmt).Quo(lockSum.Mul(sdk.NewInt(int64(remainEpochs))))
-			if amt.IsPositive() {
+			denomLockAmt := guaranteedNonzeroCoinAmountOf(lock.Coins, denom)
+			amt := coin.Amount.Mul(denomLockAmt).Quo(lockSumTimesRemainingEpochs)
+			if amt.Sign() == 1 {
 				newlyDistributedCoin := sdk.Coin{Denom: coin.Denom, Amount: amt}
 				distrCoins = distrCoins.Add(newlyDistributedCoin)
 			}
 		}
-		distrCoins = distrCoins.Sort()
+		if distrCoins.Len() > 1 {
+			// sort makes a runtime copy, due to some interesting golang details.
+			distrCoins = distrCoins.Sort()
+		}
 		if distrCoins.Empty() {
 			continue
 		}
+
 		// update the amount for that address
-		err := currResult.addLockRewards(lock.Owner, distrCoins)
+		err = currResult.addLockRewards(lock.Owner, distrCoins)
 		if err != nil {
 			return sdk.Coins{}, err
 		}
 
 		totalDistrCoins = totalDistrCoins.Add(distrCoins...)
 	}
 
-	err := k.updateGaugePostDistribute(ctx, gauge, totalDistrCoins)
+	err = k.updateGaugePostDistribute(ctx, gauge, totalDistrCoins)
 	return totalDistrCoins, err
 }
 
+// SumLocksByDenom assumes that caller is passing in locks that contain denom.
+// Source: https://github.com/osmosis-labs/osmosis/blob/v25.2.0/x/lockup/types/lock.go#L71.
+// TODO: move to x/lockup.
+func SumLocksByDenom(locks []lockuptypes.PeriodLock, denom string) (math.Int, error) {
+	sumBi := big.NewInt(0)
+	// validate the denom once, so we can avoid the expensive validate check in the hot loop.
+	err := sdk.ValidateDenom(denom)
+	if err != nil {
+		return math.Int{}, fmt.Errorf("invalid denom used internally: %s, %w", denom, err)
+	}
+	for _, lock := range locks {
+		var amt math.Int
+		// skip a 1second cumulative runtimeEq check
+		if len(lock.Coins) == 1 {
+			amt = lock.Coins[0].Amount
+		} else {
+			amt = lock.Coins.AmountOfNoDenomValidation(denom)
+		}
+		sumBi.Add(sumBi, amt.BigIntMut())
+	}
+
+	// handle overflow check here so we don't panic.
+	err = checkBigInt(sumBi)
```

Comment:
What is the reason to return `math.Int` instead of `big.Int` ?
Could you not do the subsequent calculations (`amt := coin.Amount.Mul(denomLockAmt).Quo(lockSumTimesRemainingEpochs)`) with `big.Int`, and later cast down back into `math.Int`? Since there is a division there, I have a hunch that the end result would not overflow `math.Int`.
If that's the case, then `checkBigInto` and `maxWordLen` can be removed.

Reply:
I was just thinking if it would be possible/beneficial to do all calculations with `big.Int`, and then at the end when you would have to add it to `sdk.Coin`, to downcast, and potentially check for overflow.
Doesn't matter, just thinking.